src/main/java/com/microsoft/graph/concurrency/ChunkedUploadResponseHandler.java

```diff
@@ -98,35 +107,41 @@ public void configConnection(@Nonnull final Response response) {
 		if (response.code() >= HttpResponseCode.HTTP_CLIENT_ERROR) {
 			logger.logDebug("Receiving error during upload, see detail on result error");
 
-			return new ChunkedUploadResult<UploadType>(
+			return new ChunkedUploadResult<>(
 					GraphServiceException.createFromResponse(request, null, serializer,
-						response, logger));
-		} else if (response.code() >= HttpResponseCode.HTTP_OK
+							response, logger));
+		} else if (response.code() >= HTTP_OK
 				&& response.code() < HttpResponseCode.HTTP_MULTIPLE_CHOICES) {
 			final Map<String, String> headers = responseHeadersHelper.getResponseHeadersAsMapStringString(response);
 			final String contentType = headers.get(Constants.CONTENT_TYPE_HEADER_NAME);
 			final String location = headers.get("Location");
-			if(contentType != null
-				&& contentType.contains(Constants.JSON_CONTENT_TYPE)) {
-				try (final InputStream in = new BufferedInputStream(response.body().byteStream())) {
-					final String rawJson = CoreHttpProvider.streamToString(in);
-					final IUploadSession session = serializer.deserializeObject(rawJson, uploadSessionClass);
-					if(session == null || session.getNextExpectedRanges() == null) {
-						logger.logDebug("Upload session is completed (ODSP), uploaded item returned.");
-						final UploadType uploadedItem = serializer.deserializeObject(rawJson, this.deserializeTypeClass);
-						return new ChunkedUploadResult<UploadType>(uploadedItem);
-					} else {
-						logger.logDebug("Chunk bytes has been accepted by the server.");
-						return new ChunkedUploadResult<UploadType>(session);
-					}
-				}
-			} else if(location != null) {
+			if (contentType != null
+					&& contentType.contains(Constants.JSON_CONTENT_TYPE)) {
+				return parseJsonUploadResult(response, serializer, logger);
+			} else if (location != null) {
 				logger.logDebug("Upload session is completed (Outlook), uploaded item returned.");
-				return new ChunkedUploadResult<UploadType>(this.deserializeTypeClass.getDeclaredConstructor().newInstance());
+				return new ChunkedUploadResult<>(this.deserializeTypeClass.getDeclaredConstructor().newInstance());
 			} else {
 				logger.logDebug("Upload session returned an unexpected response");
 			}
 		}
-		return new ChunkedUploadResult<UploadType>(new ClientException("Received an unexpected response from the service, response code: " + response.code(), null));
+		return new ChunkedUploadResult<>(new ClientException("Received an unexpected response from the service, response code: " + response.code(), null));
+	}
+
+	@NotNull
+	private ChunkedUploadResult<UploadType> parseJsonUploadResult(@NotNull Response response, @NotNull ISerializer serializer, @NotNull ILogger logger) throws IOException {
+		try (final InputStream in = response.body().byteStream()) {
+			final byte[] responseBytes = ByteStreams.toByteArray(in);
+			final IUploadSession session = serializer.deserializeObject(new ByteArrayInputStream(responseBytes), uploadSessionClass);
```

Comment:
wouldn't we need to close the stream? Is this stream seekable? in which case we could reuse it for the completed case

Reply:
No, not really, since closing a ByteArrayInputStream doesn't have any effect. 

Regarding the stream being seekable: I'm not an expert on the streams, but a ByteArrayInputStream can be reset() easily and that works. However, if someone called mark() on it, it won't work. Hence, from what I know, it is generally doable, but not completely reliable.

However, we can simply reusing the bytes array should do the trick, because the array is not copied or modified by the stream.

Reply:
in theory ByteArrayStream should only be a wrapper to read in the byte array providing a stream API. So its memory footprint should be very light (couple of pointers) and the close method should be a no-op as you mentioned in the other comment. I'm ok with avoiding extra complexity for a minor performance impact given that this PR drastically improves performance to begin with.
I'll let this convo open for now so other reviewers can have a look and chime in.